torch/csrc/autograd/input_buffer.cpp

```diff
@@ -86,13 +86,12 @@ static void accumulate(
     } else {
       buffer[pos] = var + old_var;
     }
+  } else if (
+      old_var.is_contiguous() && !old_var._is_zerotensor() &&
+      old_var.use_count() == 1 && old_var.storage().use_count() == 1) {
+    buffer[pos] = old_var.add_(var);
```

Comment:
I think there are a few things wrong here that need updating. In particular looking at the logic for AccumulateGrad show a lot of possible failure cases here: https://github.com/pytorch/pytorch/blob/0fa23663ccd5350469c95615ddb7d2fd2a88abe3/torch/csrc/autograd/functions/accumulate_grad.h#L56

- We need to check that grad mode is disabled. This inplace might be invalid if old_var is saved by autograd
- You need to guard for `.is_sparse_csr()` which is not covered by `.is_sparse()`
- We want is_non_overlapping_and_dense on top of is_contiguous as we really don't want weird memory overlapping Tensors to be written inplace.
- Why the special case on zero_tensor here?

Reply:
>is_non_overlapping_and_dense on top of is_contiguous

instead of is contiguous? Yeah that's reasonable. IIRC it errors out on sparse tensors though, so the logic will need to become more tortured

Reply:
FWIW I'm also updating the path where `var` is used, so there will be sparse checks before any other checks.

Reply:
> Should that be GradMode::is_enabled() or a property of old_var?

GradMode::is_enabled() because it might not be ok to modify old_var even if it doesn't require gradients.
Note that it could still be invalid when grad mode is disabled but that should be unlikely enough here that we're ok.